### PR TITLE
Set custom vertex data API (Sprites)

### LINKED ISSUE #7867: Custom vertex formats - script API support
**Is your feature request related to a problem? Please describe (REQUIRED):**
Currently the custom vertex format support is build-time only, but to gain most out of the feature we need a runtime feature that writes data into the streams from scripts (similar to constants)

**Describe the solution you'd like (REQUIRED):**

**Describe alternatives you've considered (REQUIRED):**

**Additional context (OPTIONAL):**

### PR DESCRIPTION
**Pull request by Jhonnyg**
Added support for setting and getting vertex attributes for sprites via the `go.set`, `go.get` and `go.animate` functions. Similar to shader constants, the attributes has to be specified in the material:

image

To modify the my_color attribute:

```lua
local my_color = go.get("#sprite", "my_color")
my_color.x = my_color.x + 0.5
go.set("#sprite", "my_color", my_color)
```

Or alternatively, you can use `go.animate`:

```lua
go.animate("#sprite", "my_color", go.PLAYBACK_LOOP_FORWARD, vmath.vector4(1, 0, 0, 1), go.EASING_INBACK, i))
```

Fixes #7867

### CODE REVIEW

File: engine/gamesys/src/gamesys/gamesys_private.cpp [hunk 1]
```diff
@@ -248,4 +248,345 @@ namespace dmGameSystem
...
+    {
+        for (int i = 0; i < attributes_count; ++i)
+        {
+            if (attributes[i].m_NameHash == name_hash)
+                return i;
+        }
+        return -1;
+    }
+
+    int32_t FindMaterialAttributeIndex(const DynamicAttributeInfo& info, dmhash_t name_hash)
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/gamesys_private.cpp:262]: Hm, should perhaps be called something with dynamic

File: engine/gamesys/src/gamesys/gamesys_private.cpp [hunk 2]
```diff
@@ -248,4 +248,345 @@ namespace dmGameSystem
...
+        if (use_element_index)
+        {
+            return dmGameObject::PropertyVar(values[element_index]);
+        }
+        else
+        {
+            switch(element_count)
+            {
+                case 1: return dmGameObject::PropertyVar(values[0]);
+                case 2: return dmGameObject::PropertyVar(dmVMath::Vector3(values[0], values[1], 0.0f));
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/gamesys_private.cpp:438]: We don't have a vector2

File: engine/gamesys/src/gamesys/gamesys_private.cpp [hunk 3]
```diff
@@ -248,4 +248,345 @@ namespace dmGameSystem
...
+                    dmLogError("Unable to allocate dynamic attributes, max dynamic attribute limit reached for sprites (%d)", INVALID_DYNAMIC_ATTRIBUTE_INDEX);
+                    return dmGameObject::PROPERTY_RESULT_UNSUPPORTED_VALUE;
+                }
+
+                // Put all the new indices on the free list and trim the indices list down so we don't waste too much memory
+                dynamic_attribute_free_indices.SetCapacity(DYNAMIC_ATTRIBUTE_INCREASE_COUNT);
+                for (int i = new_capacity - 1; i >= current_count; i--)
+                {
+                    dynamic_attribute_free_indices.Push(i);
+                }
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/gamesys_private.cpp:541]: we add the indices backwards since we grab new indices from the back

**Review by JCash — CHANGES_REQUESTED** [review]
looks good, but could use some simplifications I believe.
In particular the reimplementing of the pool and dynamic array functionalities

File: engine/gamesys/src/gamesys/components/comp_sprite.cpp [hunk 4]
```diff
@@ -638,18 +636,44 @@ namespace dmGameSystem
     }
 
     // Prepares the list of sprite attributes that could potentially overrides an already specified material attribute
-    static void FillSpriteAttributeInfos(SpriteAttributeInfo* material_infos, const dmGraphics::VertexAttribute* sprite_attributes, uint32_t sprite_attribute_count, SpriteAttributeInfo* sprite_infos)
+    static void FillSpriteAttributeInfos(const SpriteWorld* sprite_world, const SpriteComponent* component, SpriteAttributeInfo* material_infos, SpriteAttributeInfo* sprite_infos)
```
[the remark this answers is not in the document]
Comment (Jhonnyg) [reply engine/gamesys/src/gamesys/components/comp_sprite.cpp:639]: Yes, I think we can merge all the "attribute info" structs into a single one that lives in a graphics header or something. Particle generation is not part of gamesys and built before as a dependency to gamesys, so we need it further up in the chain. But I feel like that merge needs its own PR, I'll take a look

**Pull request merged**